### PR TITLE
Implement prefer_deferred_runtime_asserts_over_guards

### PR DESCRIPTION
**Pull request by ezyang**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #122090

Fixes https://github.com/pytorch/pytorch/issues/121749

As promised, it is pretty easy.

Signed-off-by: Edward Z. Yang <ezyang@meta.com>

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @chenyang78 @kadeng @chauhang

### CODE REVIEW

File: torch/_dynamo/output_graph.py [hunk 1]
```diff
@@ -1402,6 +1402,33 @@ def insert_deferred_runtime_asserts(self, root, name) -> None:
...
+                    ).node
+                    self.graph.call_function(
+                        torch.ops.aten._assert_scalar.default,
+                        # TODO: use ra.msg here, but it's pretty
+                        # useless right now
+                        (
+                            res,
+                            f"Deferred runtime assertion failed {ra.expr}",
+                        ),
+                    )
```
Comment (ezyang) [inline torch/_dynamo/output_graph.py:1430]: Code motion only

File: torch/_dynamo/output_graph.py [hunk 2]
```diff
@@ -1402,6 +1402,33 @@ def insert_deferred_runtime_asserts(self, root, name) -> None:
...
         log.debug("needed_symbols = %s", needed_symbols)
 
+        def add_runtime_asserts(ras):
+            for ra in ras:
+                log.debug("inserting runtime assert %s", ra.expr)
+                # Need to process ALL free symbols, not just unbacked ones
+                fvs = free_symbols(ra.expr)
+                missing = fvs - symbol_to_proxy.keys()
+                if missing:
+                    i1 = sorted(missing)[0]
```
Comment (Skylion007) [inline torch/_dynamo/output_graph.py:1412]:
```suggestion
                    i1 = min(missing)
```
Isn't that equivalent and more efficient?

File: torch/fx/experimental/symbolic_shapes.py [hunk 3]
```diff
@@ -4284,7 +4293,10 @@ def defer_runtime_assert(self, orig_expr: "sympy.Expr", msg, fx_node=None):
             ra = RuntimeAssert(expr, msg, stack)
             # TODO: Do this in a way that is less janky than int(s.name[1:])
             cands = sorted([s for s in expr.free_symbols if s.name.startswith("u")], key=lambda s: int(s.name[1:]))
-            self.deferred_runtime_asserts.setdefault(cands[-1], []).append(ra)
+            # Is None when prefer_deferred_runtime_asserts_over_guards=True
+            # and the guard in question has no unbacked SymInts in front
```
Comment (lezcano) [inline torch/fx/experimental/symbolic_shapes.py:4297]: add an assert

**Review by lezcano — COMMENTED** [review]
Do you mind setting the default to true and run CI on it, see if CI is happy with the new path?

**Comment by lezcano** [comment]
Sure, but even then it'd be good to see that all the errors are in paths where an assertion fails.

**Comment by ezyang** [comment]
Looking over the CI results, some thoughts:

* torch._dynamo.exc.InternalTorchDynamoError: type object 'PythonReferenceAnalysis' has no attribute 'round', simple enough, I suppose. `python test/inductor/test_torchinductor.py -k test_builtins_round_cuda` . Related round brokeness `PYTORCH_TEST_WITH_DYNAMO=1 python test/nn/test_pooling.py -k test_pooling_max_nhwc_cpu_float32` "InternalTorchDynamoError: must be real number, not Proxy"
* Some `guard_equals` tests in Inductor are now failing asserts. Example `python test/inductor/test_torchinductor_codegen_dynamic_shapes.py -k test_alexnet_prefix_dynamic_shapes_cpu`
* Interesting variable referenced before assignment `python test/dynamo/test_aot_autograd.py -k test_aot_sequence_nr`
* Triggers add_runtime_asserts "cannot determine truth value of Relational" `python test/dynamo/test_dynamic_shapes.py -k test_bigbird_unsqueeze_inplace_dynamic_shapes` ; https://github.com/pytorch/pytorch/pull/121079 is supposed to fix this

**Review by lezcano — APPROVED** [review]

**Pull request closed without merging**